torch/nn/utils/parametrizations.py

```diff
@@ -1,10 +1,273 @@
+        # * 0 0
+        # * * 0
+        # which, for complex matrices, give n(n-1) (real) parameters. Now, you need n^2 parameters
+        # to parametrize the unitary matrices. Saving tau on its own does not work either, because
+        # not every combination of `(A, tau)` gives a unitary matrix, meaning that if we optimise
+        # them as independent tensors we would not maintain the constraint
+        # An equivalent reasoning holds for rectangular # matrices
+        if weight.is_complex() and parametrization == "householder":
+            raise ValueError("The householder parametrization does not support complex tensors.")
+
+        # We could implement this for 1-dim tensors as the maps on the sphere
+        # but I believe it'd bite more people than it'd help
+        if weight.ndim < 2:
+            raise ValueError("Expected a matrix or batch of matrices. "
+                             f"Got a tensor of {weight.ndim} dimensions.")
+
+        self.shape = weight.shape
+        self.parametrization = parametrization
+        if extra_memory:
+            self.register_buffer("base", None)
+
+    def forward(self, X: torch.Tensor) -> torch.Tensor:
+        if X.shape != self.shape:
+            raise ValueError(f"Expected a matrix or batch of matrices of shape {self.shape}. "
+                             f"Got a tensor of shape {X.shape}.")
+
+        n, k = X.size(-2), X.size(-1)
+        transposed = n < k
+        if transposed:
+            X = X.transpose(-2, -1)
+            n, k = k, n
+        # Here n > k and X is a tall matrix
+        if self.parametrization == "matrix_exp" or self.parametrization == "cayley":
+            # We just need n x k - k(k-1)/2 parameters
+            X = X.tril()
+            if n != k:
+                # Embed into a square matrix
+                X = torch.cat([X, X.new_zeros(n, n - k).expand(*X.shape[:-2], -1, -1)], dim=-1)
+            A = X - X.transpose(-2, -1).conj()
+            # A is skew-symmetric (or skew-hermitian)
+            if self.parametrization == "matrix_exp":
+                Q = torch.matrix_exp(A)
+            elif self.parametrization == "cayley":
+                # Computes the Cayley retraction (I+A/2)(I-A/2)^{-1}
+                Id = torch.eye(n, dtype=A.dtype, device=A.device)
+                Q = torch.linalg.solve(torch.add(Id, A, alpha=-0.5), torch.add(Id, A, alpha=0.5))
+            # Q is now orthogonal (or unitary) of size (..., n, n)
+            if n != k:
+                Q = Q[..., :k]
+            # Q is now the size of the original tensor
+        else:
+            # X is real here, as we do not support householder with complex numbers
+            A = X.tril(diagonal=-1)
+            tau = 2. / (1. + (A * A).sum(dim=-2))
+            Q = torch.linalg.householder_product(A, tau)
+            # The diagonal of X is 1's and -1's
+            # We do not want to differentiate through this or update the diagonal of X hence the casting
```

Comment:
Why don't we want to differentiate through this?

Reply:
We do not want to update the `1`s and `-1`s in the diagonal, we want to treat them as constants (we make sure we did not update them in the tests). Is there a better way to do this than the one I wrote? I was not sure how to write this bit.

Reply:
With torch.no_grad() is the usual way to go to prevent autograd from tracking an op.
Or if you want a given Tensor not to be considered, whatever its history, you can use .detach()

Should this behavior be mentioned in the doc?

Reply:
No, it is fine not to mention it. This is expected. This tensor could be saved as a buffer, but I save it in the diagonal to keep things more memory efficient.